README.md
# Pyostra Package
Custom terminal colored logging system developed for the [Advanced NFTs Generator](https://github.com/ostra-project/Advanced-NFTs-Generator) and maintained for future projects.

[PyPI for releases](https://pypi.org/project/pyostra/) | [Github for source](https://github.com/yoratoni/pyostra)


## Installation
Developed for Python 3.9, normally supports CPython 2.7 => 3.10 | Pypy 2.7 and 3.6.

Only requirement other than the standard library is [Colorama](https://github.com/tartley/colorama), this package have the same Python / platform requirements.

```
pip install pyostra
```


## Description
Pyostra is the main central library used for my own projects such as the [Advanced NFTs Generator](https://github.com/ostra-project/Advanced-NFTs-Generator). This library allows custom colored and formatted outputs in the Terminal.
The goal is to simplify all the interactions with the terminal by the developers such as try / except errors etc..

Note that this library will later include many other general functions, it is not specifically made for Terminal outputs only but to centralize all of the general functions that I'm using in my projects instead of copying the code everytime. I still decided to make it public so people can use it too if they want.


## LogTypes class
This class contains all the log types for the `pyprint()` function.

It contains all of these log types by default:
- CRITICAL
- ERROR
- WARN
- SUCCESS
- SILENT
- READY
- DATA
- INFO

A log type can also be created by using the `new_log_type()` function, it returns a dict with the right format.

The dict used for a log type is pretty simply:
```py
DATA = {
    # The default name used by the log messages
    'longName': 'DATA',
    
    # Check the set_short() function doc below to understand
    'shortName': 'DATA',
    
    # The color of the log type at the start
    'typeColor': Fore.LIGHTCYAN_EX,
    
    # The color of the log message
    'msgColor': Fore.LIGHTCYAN_EX
}
```


## pyprint() function
Default format of a Pyprint log, using the default log types:
```py
from pyostra import LogTypes, pyprint

def main():
    pyprint(LogTypes.CRITICAL, 'THIS IS A TEST')
    pyprint(LogTypes.ERROR, 'THIS IS A TEST')
    pyprint(LogTypes.WARN, 'THIS IS A TEST')
    pyprint(LogTypes.SUCCESS, 'THIS IS A TEST')
    pyprint(LogTypes.SILENT, 'THIS IS A TEST')
    pyprint(LogTypes.READY, 'THIS IS A TEST')
    pyprint(LogTypes.DATA, 'THIS IS A TEST')
    pyprint(LogTypes.INFO, 'THIS IS A TEST')
    
main()
```
Notice that `CRITICAL`, `ERROR` and `WARN` log types also returns the calling function name,
it can be disabled by using the optional `disable_function_name` argument for every log where you don't want the calling function name to show.

![](https://github.com/yoratoni/pyostra/blob/main/documentation/default_output.png "pyprint function example")


## set_short() function
This function allows you to reduce the log type length to 4 chars, it can be used if you have a lot of logs at the same time, for better readability. It uses the `shortName` string key inside a log type dict.
```py
import pyostra

pyostra.set_short(True)
```
If not clear, all the correspondances with the full log types can be found inside the pyprint/set_short function docstring.

![](https://github.com/yoratoni/pyostra/blob/main/documentation/set_short.png "set_short function example")


## set_verbose() function
This function allows to you to reduce the amount of logs inside the Terminal by ignoring some log types.
```py
from pyostra import LogTypes, pyprint, set_verbose

set_verbose(False)

def main():
    pyprint(LogTypes.CRITICAL, 'THIS IS A TEST')
    pyprint(LogTypes.ERROR, 'THIS IS A TEST')
    pyprint(LogTypes.WARN, 'THIS IS A TEST')
    pyprint(LogTypes.SUCCESS, 'THIS IS A TEST')
    pyprint(LogTypes.SILENT, 'THIS IS A TEST')  # IGNORED
    pyprint(LogTypes.READY, 'THIS IS A TEST')  # IGNORED
    pyprint(LogTypes.DATA, 'THIS IS A TEST')  # IGNORED
    
    # IGNORED BY DEFAULT BUT FORCED WITH THE force_verbose ARG
    pyprint(LogTypes.INFO, 'THIS IS A TEST', force_verbose=True)
    
main()
```
By setting the verbose to False, only `CRITICAL`, `ERROR`, `WARN` and `SUCCESS` log types will be printed, the other ones are ignored.

![](https://github.com/yoratoni/pyostra/blob/main/documentation/set_verbose.png "set_verbose function example")

In addition to that, the optional argument called `force_verbose` inside the `pyprint()` function allows to you print other log types even when the verbose debugging is turned off.


## new_section() function
This simple function allows you to separate different types of log by printing a title with a bunch of separators.

This function contains only optional args:
- section_title (str, optional): The title of the current section.
- skip_line (bool, optional): If True, skip a line before the title.
- section_title_color (str, optional): Use Colorama for that.
- separator_char (str, optional): The character used by the section title.
- separators_amount (int, optional): The amount of separators for the title.

```py
from pyostra import LogTypes, new_section, pyprint

def main():
    new_section('TEST SECTION', separators_amount=10)
    pyprint(LogTypes.CRITICAL, 'TEST')
```

![](https://github.com/yoratoni/pyostra/blob/main/documentation/new_section.png "new_section function example")


## extime() function
This function uses the `time` core library, specifically the `time.perf_counter_ns()` function to measure the performances.
It prints the final message into the Terminal by using the `pyprint()` function (`LogTypes.SUCCESS` type).

This function contains a bunch of args:
- timer_name (str): Name of the timer.
- perf_counter_ns_value (int): Using time.perf_counter_ns() to get the starting value of the timer.
- multiply_timer (int, optional): Multiply the time by a value (To estimate time for x iterations).
- approximated_value (bool, optional): If True, adds the "~" character to show that it's an approximation.

The `multiply_timer` and the `approximated_value` args can be used to estimate the time that it takes to execute a function.
As an example, inside the [Advanced NFTs Generator](https://github.com/ostra-project/Advanced-NFTs-Generator), I'm generating an NFT 5 times in a row (for a better estimation) to measure the time that it takes, then, I set the `multiply_timer` value to the amount of NFTs that needs to be generated (divided by 5 as the measured time is for 5 NFTs). Then I set the `approximated_value` value to True, it adds the little `~` char to show that it is an estimation.

```py
from pyostra import extime
import time

timer = time.perf_counter_ns()

a = 0
for i in range(10000):
    a += 1

extime('Loop execution time', timer)
```

![](https://github.com/yoratoni/pyostra/blob/main/documentation/extime.png "new_section function example")


CHANGELOG.md
## Changelog

### v0.1.4 (03/06/2022):
- Complete refactoring.
- README doc fully written.
- Main pyprint functions added.
- Now supports short log types.
- Logging system using dictionaries.
- Functions docstring added.
- Extime function refactored.
- Arguments & docstring refactoring.

### v0.1.1 (03/05/2022):
- Package setup updated.
- Required packages updated.
- General refactoring.

### v0.0.1 (03/05/2022):
- First release of the pyostra package.


test.py
from pyostra import LogTypes, pyprint, set_short

set_short(True)

def main():
    pyprint(LogTypes.CRITICAL, 'THIS IS A TEST')
    pyprint(LogTypes.ERROR, 'THIS IS A TEST')
    pyprint(LogTypes.WARN, 'THIS IS A TEST')
    pyprint(LogTypes.SUCCESS, 'THIS IS A TEST')
    pyprint(LogTypes.SILENT, 'THIS IS A TEST')
    pyprint(LogTypes.READY, 'THIS IS A TEST')
    pyprint(LogTypes.DATA, 'THIS IS A TEST')
    pyprint(LogTypes.INFO, 'THIS IS A TEST')
    
main()

core/debug.py
from colorama import Style, Fore

import inspect
import time


class DebugSettings:
    # If true, print all the available logs,
    # Else, just the types defined in 'forced_types'
    verbose_debugging = True
    
    # The section title, to separate the logs
    section_separators = '-' * 35
    section_color = Fore.LIGHTBLUE_EX
    section_title = 'PYPRINT SECTION'
    
    # List of debug log types and their colors in a dict
    all_types = {
        'INFO': Fore.LIGHTBLUE_EX,
        'DATA': Fore.CYAN,
        'WARNING': Fore.YELLOW,
        'ERROR': Fore.LIGHTRED_EX,
        'SUCCESS': Fore.LIGHTGREEN_EX
    }
    
    # Debug log types that are always printed (even when verbose debugging is turned off)
    forced_types = ['WARNING', 'ERROR', 'SUCCESS']
    
    # Debug log types that shows the name of the function that calls pyprint()
    show_function_name_types = ['WARNING', 'ERROR']
    
    # Status codes used for general errors
    # I usually use modified HTTP Status Codes
    # Simply use 'log_or_status' with 'ST_' and the code after it (log_or_status = 'ST_42')
    status_codes = {
        42: 'Pyprint Status Example'
    }


class Logger:
    __is_first_print = True
    
    
    @staticmethod
    def __show_section_title():
        '''Show the section title (printed only once).
        It separates the normal console logs and pyprint() ones.
        '''

        separated_title = f'{DebugSettings.section_separators}{DebugSettings.section_title}{DebugSettings.section_separators}'
        print(f'\n{DebugSettings.section_color}{separated_title}{Style.RESET_ALL}')
        

    @staticmethod
    def pyprint(
        log_type: str,
        log_or_status: str,
        same_line: bool = False,
        disable_function_name: bool = False
    ):
        '''Debug Mode formatted print statements.
        
        Supported message types:
        - INFO -> Light blue
        - DATA -> Cyan
        - WARNING -> Yellow
        - ERROR -> Light red
        - SUCCESS -> Light green

        Args:
            log_type (str): Type of the log (Unsupported title returns white colored log).
            log_or_status_code (str): Printed log message or a status code if 'ST_000'
                where '000' is the status code integer corresponding to the one in the settings.
            same_line (bool, optional): Print on the same line as before.
            disable_function_name (bool, optional): Allows to specifically disable the printing
                of the calling function (Cancels the 'show_function_name_types' specifically for a print).
        '''

        if (not DebugSettings.verbose_debugging and log_type in DebugSettings.forced_types) or DebugSettings.verbose_debugging:
            # Show the section title
            if Logger.__is_first_print:
                Logger.__show_section_title()
                Logger.__is_first_print = False
        
            # Get the color of the log
            if log_type in DebugSettings.all_types.keys():
                color = DebugSettings.all_types[log_type]
            else:
                color = Fore.WHITE
                
            # Check if it's a status code
            if not log_or_status.startswith('ST_'):
                
                # Basic normal output
                if (log_type not in DebugSettings.show_function_name_types) or disable_function_name:
                    output = f'[{log_type}] {log_or_status}'
                else:
                    # Get the function that calls pyprint
                    cur_frame = inspect.currentframe()
                    out_frame = inspect.getouterframes(cur_frame, 2)
                    output = f'[{log_type}] {out_frame[1][3]}(): {log_or_status}'
            
            # Status code case   
            else:
                if log_or_status in DebugSettings.status_codes:
                    output = f'[{log_or_status}] {DebugSettings.status_codes[int(log_or_status[3:])]}'
                    
                # Non-existing status code catch
                else:
                    color = Fore.LIGHTRED_EX
                    output = f'[PYPRINT_ERROR] Wrong status code, "{log_or_status}" does not exists'

            # Same line handling
            if same_line:
                print_end = '\r'
            else:
                print_end = None
            
            # Final output
            print(f'{color}{output}{Style.RESET_ALL}', end=print_end)
            
            
    @staticmethod
    def extime(
        name: str,
        timer: int,
        multiply_timer: int = 1,
        print_msg: bool = True,
        approximated_value: bool = False,
    ) -> str:
        '''Automatic timer format (ns, µs, ms, s and mins units).
        
        Args:
            name (str): Name of the timer.
            timer (int): Using time.perf_counter_ns() to get the starting point of the timer.
            multiply_timer (int, optional): Multiply the time by a value (To estimate time).
            print_msg (bool, optional): If True, it also prints the formatted timer message.
            approximated_value (bool, optional): If True, adds the "~" character for an approximation.
            
        Returns:
            str: The formatted timer message
        '''

        timer = (time.perf_counter_ns() - timer) * multiply_timer
        
        units = ['ns', 'µs', 'ms', 's', ' mins']
        powers = [10**3, 10**6, 10**9]
        res = 0
        i = 0
        
        if timer < powers[0]:
            res = timer
        elif powers[0] <= timer < powers[1]:
            res = round(timer / powers[0])
            i = 1
        elif powers[1] <= timer < powers[2]:
            res = round(timer / powers[1])
            i = 2
        elif powers[2] <= timer:
            res = timer / powers[2]
            i = 3
            
            # Using minutes instead
            if res > 120:
                res = round(res / 60)
                i = 4
        
        res = round(res, 2)
        
        if approximated_value:
            output = f'{name}: ~{res}{units[i]}'
        else:
            output = f'{name}: {res}{units[i]}'
        
        if print_msg:
            Logger.pyprint('SUCCESS', output)
        
        return output


pyostra/__init__.py
'''Encapsulation of all the Ostra python environment main functions.

This module provides a custom formatted logging system,
it also implements a formatted performance output system (extime).

set_verbose(): if True, prints all the available log messages.
In the other case, prints only important messages depending on the log type (Such as Warn, Error and Critical).

set_short(): As the log type is printed at the start of every log,
If set to True, it reduces the log types to 4 chars only, making it more readable if you have a lot of printing.

new_log_type(): Defines a new type using Colorama colors (it returns a dict that can be used as an arg for pyprint())

new_section(): Used to separate different types of logs.

pyprint(): The main custom formatted print statements.

extime(): Using time.perf_counter_ns(), returns a formatted value of the time.
'''

from pyostra.logger import LogTypes
from pyostra.logger import set_verbose
from pyostra.logger import set_short
from pyostra.logger import new_log_type
from pyostra.logger import new_section
from pyostra.logger import pyprint
from pyostra.logger import extime


settings/parameters.py
class GlobalSettings:
    '''The debugging settings for the custom Logger class.
    '''
    
    # Debugging
    dist_mode = False  # If True, remove all the console messages, even forced ones
    verbose_debugging = True  # Print a lot more data about the NFTs
    debug_types = ['INFO', 'DATA', 'WARN', 'ERRO', 'SUCCESS']  # List of debug message types
    
    # Status codes used for general errors
    # I usually use modified HTTP Status Codes
    # Simply use the parameter 'status_code' with the code ID inside pyprint
    status_codes = {
        42: 'Pyprint Status Example',
        
        100: 'Continue',
        102: 'Processing',
        
        200: 'OK',
        201: 'Created',
        202: 'Accepted',
        
        400: 'Bad Request',
        401: 'Unauthorized',
        403: 'Forbidden',
        404: 'Not Found',
        405: 'Method Not Allowed',
        409: 'Conflict',
        429: 'Too Many Requests',
        
        500: 'Internal Server Error',
        501: 'Not Implemented',
        503: 'Service Unavailable',
        511: 'Network Authentication Required',
        599: 'Network Connect Timeout Error'
    }

pyostra/logger.py
'''
Made by Yoratoni.
Pyostra is licensed under the BSD 3-Clause "New" or "Revised" License.

All rights reserved to Jonathan Hartley for the Colorama library.
GITHUB: https://github.com/tartley/colorama
LICENSE: https://github.com/tartley/colorama/blob/master/LICENSE.txt
'''



from inspect import currentframe, getouterframes
from colorama import Style, Fore, init
from typing import Any
import time


class LogTypes():
    '''Contains all the available log types for the pyprint() function.
    These types comes with a short version, every output type will have the same length (4 chars).
    
    Available types:
        - [CRIT] CRITICAL
        - [ERRO] ERROR
        - [WARN] WARN
        - [SCES] SUCCESS
        - [SILT] SILENT
        - [REDY] READY
        - [DATA] DATA
        - [INFO] INFO
    '''

    CRITICAL = {'longName': 'CRITICAL', 'shortName': 'CRIT', 'typeColor': Fore.LIGHTRED_EX,     'msgColor': Fore.LIGHTRED_EX}
    ERROR    = {'longName': 'ERROR',    'shortName': 'ERRO', 'typeColor': Fore.RED,             'msgColor': Fore.RED}
    WARN     = {'longName': 'WARN',     'shortName': 'WARN', 'typeColor': Fore.YELLOW,          'msgColor': Fore.YELLOW}
    SUCCESS  = {'longName': 'SUCCESS',  'shortName': 'SCES', 'typeColor': Fore.LIGHTGREEN_EX,   'msgColor': Fore.LIGHTGREEN_EX}
    SILENT   = {'longName': 'SILENT',   'shortName': 'SILT', 'typeColor': Fore.LIGHTBLACK_EX,   'msgColor': Fore.LIGHTBLACK_EX}
    READY    = {'longName': 'READY',    'shortName': 'REDY', 'typeColor': Fore.LIGHTMAGENTA_EX, 'msgColor': Fore.LIGHTMAGENTA_EX}
    DATA     = {'longName': 'DATA',     'shortName': 'DATA', 'typeColor': Fore.LIGHTCYAN_EX,    'msgColor': Fore.LIGHTCYAN_EX}
    INFO     = {'longName': 'INFO',     'shortName': 'INFO', 'typeColor': Fore.LIGHTBLUE_EX,    'msgColor': Fore.LIGHTBLUE_EX}
    

class __Settings:
    # Verbose mode for printing
    VERBOSE_DEBUGGING = True
    
    # Short mode for printing (log types of 4 chars)
    SHORT_DEBUGGING = False

    # Type functionalities
    FORCED_TYPES = [
        LogTypes.SUCCESS,
        LogTypes.WARN,
        LogTypes.ERROR,
        LogTypes.CRITICAL
    ]
    SHOW_FUNCTION_NAME_TYPES = [
        LogTypes.WARN,
        LogTypes.ERROR,
        LogTypes.CRITICAL
    ]


def set_verbose(state: bool):
    '''Change the state of the verbose debugging (Set to True by default).
    Verbose debugging simply means that all available log types are printed into the console.
    
    If turned off, only main log types will be printed:
        - CRITICAL
        - ERROR
        - WARN
        - SUCCESS.
    '''
    
    __Settings.VERBOSE_DEBUGGING = state
    
    
def set_short(state: bool):
    '''Change the state of the short debugging (Set to False by default).
    Short debugging simply means that log types in the log only contains 4 chars.
    
    Examples:
        - "[CRITICAL] Message.." -> "[CRIT] Message.."
        - "[SILENT] Message.." -> "[SILT] Message.."
        
    This allows to print messages with the same starting point,
    it's useful if you have a lot of print statements, it makes it more readable.
    '''
    
    __Settings.SHORT_DEBUGGING = state


def new_log_type(
    long_name: str,
    short_name: str,
    type_color: str = Fore.WHITE,
    msg_color: str = Fore.WHITE
) -> dict:
    '''Allows to define a new log type as a dict,
    The value is returned and can be saved into a type var.
    
    Args:
        long_name (str): The long version of the log type name (Used by default).
        short_name (str): The short version of the log type name (if set_short(True)).
        msg_color (Back.COLOR / Fore.COLOR, optional): Color of the message itself.
        type_color (Back.COLOR / Fore.COLOR, optional): Color of the type that comes before the msg.
    
    Returns:
        dict: A logtype dict.
    '''
    
    return {
        'longName': long_name,
        'shortName': short_name,
        'typeColor': type_color,
        'msgColor': msg_color
    }
    


def new_section(
    section_title: str = 'PYOSTRA DEBUGGING',
    skip_line: bool = True,
    section_title_color: str = Fore.LIGHTBLUE_EX,
    separator_char: str = '=',
    separators_amount: int = 35,
):
    '''Show a new section title, to separate different part during the debugging.
    Sections allows to separate different types of console logs.
    
    Example:
        ============= NEW SECTION =============
    
    Note:
        All the parameters are optional and can be used to customize the section title.
    
    Args:
        section_title (str, optional): The title of the current section.
        skip_line (bool, optional): If True, skip a line before the title.
        section_title_color (str, optional): Use Colorama for that (Fore.LIGHTBLUE_EX for example).
        separator_char (str, optional): The character used by the section title (see the example).
        separators_amount (int, optional): The amount of separators for the title.
    '''
    
    separators = separator_char * separators_amount
    separated_title = f'{separators} {section_title} {separators}'
    output = f'{section_title_color}{separated_title}{Style.RESET_ALL}'
    
    if skip_line:
        output = f'\n{output}'
    print(output)


def __internal_error(log_msg: str):
    print(f'{Fore.LIGHTRED_EX}[PYOSTRA_ERROR] {log_msg}{Style.RESET_ALL}')


def pyprint(
    log_type: dict,
    log_msg: Any,
    disable_function_name: bool = False,
    on_same_line: bool = False,
    force_verbose: bool = False
):
    '''Pyostra formatted print statements.
    
    The "log_type" argument needs a dict that comes from the "LogTypes" class.
    But you can also define your own log type by using the new_log_type() function.
    
    Supported log types by default:
        - [CRIT] CRITICAL
        - [ERRO] ERROR
        - [WARN] WARN
        - [SCES] SUCCESS
        - [SILT] SILENT
        - [REDY] READY
        - [DATA] DATA
        - [INFO] INFO
        
    Note:
        By default, critical errors and normal errors also prints the name of the calling function,
        "disable_function_name" simply allows you to bypass/disable that for a specific log.
        
    Args:
        log_type (str): Type of the log (Unsupported log type returns white colored log).
        log_msg (str): The main message of the log.
        disable_function_name (bool, optional): Allows to specifically disable the printing
            of the caller's name for errors & warnings.
        on_same_line (bool, optional): Print on the same line as before.
        force_verbose (bool, optional): Force the log inside the terminal,
            not depending on the log type when verbose debugging is turned off.
    '''
    
    # Verbose debugging, forced log type or forced verbose
    if (__Settings.VERBOSE_DEBUGGING
        or (not __Settings.VERBOSE_DEBUGGING and log_type in __Settings.FORCED_TYPES)
        or force_verbose
    ):
        # Validate the log_type dict
        if type(log_type) == dict:
            if set(('longName', 'shortName', 'typeColor', 'msgColor')) <= log_type.keys():
                long_name = log_type['longName']
                short_name = log_type['shortName']
                type_color = log_type['typeColor']
                msg_color = log_type['msgColor']
            else:
                __internal_error(f'Invalid log type key error: {log_type}')
                return
        else:
            __internal_error(f'Specified log type is not a dict: {log_type}')
            return    
        
        # Short log type debugging
        if __Settings.SHORT_DEBUGGING:
            log_type_name = short_name
        else:
            log_type_name = long_name
            
        # Show the caller's name
        if log_type not in __Settings.SHOW_FUNCTION_NAME_TYPES or disable_function_name:
            function_name = ''
        else:
            cur_frame = currentframe()
            out_frame = getouterframes(cur_frame, 2)
            name = out_frame[1][3]
            
            if name != '<module>':
                name = f'{name}():'
            else:
                name = 'Module:'
                
            function_name = f' {name}'
            
        # Print on the previous line
        if on_same_line:
            end_char = '\r'
        else:
            end_char = None
        
        # Final print statement
        init(True)
        log_title = f'{type_color}[{log_type_name}]'
        print(f'{log_title}{msg_color}{function_name} {log_msg}', end=end_char)


def extime(
    timer_name: str,
    perf_counter_ns_value: int,
    multiply_timer: int = 1,
    approximated_value: bool = False,
):
    '''Automatic timer format (ns, µs, ms, s and mins units).
    
    Args:
        timer_name (str): Name of the timer.
        perf_counter_ns_value (int): Using time.perf_counter_ns() to get the starting value of the timer.
        multiply_timer (int, optional): Multiply the time by a value (To estimate time for x iterations).
        approximated_value (bool, optional): If True, adds the "~" character to show that it's an approximation.
        
    Returns:
        str: The formatted timer message
    '''

    timer = (time.perf_counter_ns() - perf_counter_ns_value) * multiply_timer
    
    units = ['ns', 'µs', 'ms', 's', ' mins']
    powers = [10**3, 10**6, 10**9]
    res = 0
    i = 0
    
    if timer < powers[0]:
        res = timer
    elif powers[0] <= timer < powers[1]:
        res = round(timer / powers[0])
        i = 1
    elif powers[1] <= timer < powers[2]:
        res = round(timer / powers[1])
        i = 2
    elif powers[2] <= timer:
        res = timer / powers[2]
        i = 3
        
        # Using minutes instead
        if res > 120:
            res = round(res / 60)
            i = 4
    
    res = round(res, 2)
    
    if approximated_value:
        output = f'{timer_name}: ~{res}{units[i]}'
    else:
        output = f'{timer_name}: {res}{units[i]}'
    
    pyprint(LogTypes.SUCCESS, output)
